File: MATTHEW/leetcode/AIcoding/5.Qwen-3/c_solutions/913-cat-and-mouse.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MOUSE_WIN 1
#define CAT_WIN 2
#define DRAW 0
#define UNKNOWN -1

int dp[101][101][2];
int g[101][101];

int solve(int mouse, int cat, int turn) {
    if (mouse == 0) return MOUSE_WIN;
    if (mouse == cat) return CAT_WIN;
    if (turn == 200) return DRAW;

    if (dp[mouse][cat][turn % 2] != UNKNOWN) {
        return dp[mouse][cat][turn % 2];
    }

    int is_mouse_turn = turn % 2 == 0;
    int result = is_mouse_turn ? CAT_WIN : MOUSE_WIN;

    for (int next = 0; next < 101; next++) {
        if (g[mouse][next] == 0) continue;

        if (is_mouse_turn) {
            int next_result = solve(next, cat, turn + 1);
            if (next_result == MOUSE_WIN) {
                result = MOUSE_WIN;
                break;
            } else if (next_result == DRAW) {
                result = DRAW;
            }
        } else {
            int next_result = solve(mouse, next, turn + 1);
            if (next_result == CAT_WIN) {
                result = CAT_WIN;
                break;
            } else if (next_result == DRAW) {
                result = DRAW;
            }
        }
    }

    dp[mouse][cat][turn % 2] = result;
    return result;
}

int catMouseGame(int* mouse, int mouseSize, int* cat, int catSize, int** graph, int graphSize, int* graphColSize) {
    for (int i = 0; i < 101; i++) {
        for (int j = 0; j < 101; j++) {
            for (int k = 0; k < 2; k++) {
                dp[i][j][k] = UNKNOWN;
            }
        }
    }

    for (int i = 0; i < graphSize; i++) {
        for (int j = 0; j < graphColSize[i]; j++) {
            g[i][graph[i][j]] = 1;
        }
    }

    return solve(*mouse, *cat, 0);
}

File: leetcode/AIcoding/5.Qwen-3/javascript_solutions/913-cat-and-mouse.js
function catMouseGame(mouse, cat, move) {
    const MOUSE_WIN = 0;
    const CAT_WIN = 1;
    const DRAW = 2;
    const MAX_MOVES = 100;

    const graph = {};
    for (let i = 0; i < mouse * 2; i++) {
        graph[i] = [];
    }

    for (let i = 0; i < mouse * 2; i++) {
        for (let j = 0; j < cat * 2; j++) {
            for (let k = 0; k < 2; k++) {
                graph[i * 2 + j * 2 + k] = [];
            }
        }
    }

    for (let i = 0; i < mouse * 2; i++) {
        for (let j = 0; j < cat * 2; j++) {
            for (let k = 0; k < 2; k++) {
                const pos = i * 2 + j * 2 + k;
                for (let next = 0; next < mouse * 2; next++) {
                    if (next !== i) {
                        graph[pos].push(next * 2 + j * 2 + 1 - k);
                    }
                }
                for (let next = 0; next < cat * 2; next++) {
                    if (next !== j) {
                        graph[pos].push(i * 2 + next * 2 + 1 - k);
                    }
                }
            }
        }
    }

    const memo = {};
    const dfs = (m, c, turn) => {
        const key = m * 2 + c * 2 + turn;
        if (memo[key] !== undefined) {
            return memo[key];
        }

        if (m === 0) return MOUSE_WIN;
        if (m === cat * 2) return CAT_WIN;
        if (turn === MAX_MOVES) return DRAW;

        const isMouseTurn = turn % 2 === 0;
        let result = isMouseTurn ? CAT_WIN : MOUSE_WIN;

        for (const next of graph[m * 2 + c * 2 + turn]) {
            const res = dfs(next >> 1, next & 1, turn + 1);
            if (isMouseTurn) {
                if (res === MOUSE_WIN) {
                    result = MOUSE_WIN;
                    break;
                } else if (res === DRAW) {
                    result = DRAW;
                }
            } else {
                if (res === CAT_WIN) {
                    result = CAT_WIN;
                    break;
                } else if (res === DRAW) {
                    result = DRAW;
                }
            }
        }

        memo[key] = result;
        return result;
    };

    return dfs(mouse, cat, 0);
}